### DISCUSSION TITLE
Prisma Migrate: Deploy Migration with Docker

### DISCUSSION THREAD

**Discussion by giscus** [notiz-dev/notiz#190] [General]
# Prisma Migrate: Deploy Migration with Docker

Perform database migration with Prisma Migrate using Docker

https://notiz.dev/blog/prisma-migrate-deploy-with-docker

**Comment by ratasorin** [comment]
Beautiful work !!!

**Comment by anuragk15** [comment]
Excellent tutorial.
We don't need to run prisma generate anywhere?

**Reply by marcjulian** [reply]
Thank you @anuragk15!

`@primsa/generate` has it's own `postinstall` script, which will generate the client for the first `npm i`. Thus you don't need to trigger prisma generate separately.

**Reply by madaster97** [reply]
Hi Marc, do you think there's merit to running generate manually after the npm i? That way you're not re-installing all dependencies every time you update your schema. So something like this:
```
COPY package.json package-lock.json ./
RUN npm ci

COPY prisma ./prisma/
RUN npx prisma generate
```

**Comment by Yasir5247** [comment]
```
#----------------------- base ------------------------#
FROM node:12-alpine AS base

WORKDIR /app

#----------------------- builder ---------------------#

FROM base AS builder

COPY package*.json ./
COPY tsconfig*.json ./
COPY prisma ./prisma/

RUN npm install --only=production

RUN cp -R node_modules node_modules_production

RUN npm install

COPY ./src ./src

RUN npm run build

#----------------------- Tests -----------------------#

# FROM builder AS tests
# RUN npm run test

#----------------------- Release ---------------------#

FROM builder AS release

COPY --from=builder /app/node_modules_production ./node_modules
COPY --from=builder /app/build ./build
COPY --from=builder /app/prisma ./prisma

USER node

EXPOSE 8080

CMD ["npm", "run", "start:migrate:prod"]

```

"scripts": {
    "postinstall": "generate:prisma && generate:nexus",
    "prebuild": "rimraf build",
    "build": "tsc",
    "start:prod": "node build/src/server.js",
    "start:migrate:prod": "prisma migrate deploy && npm run start:prod",
    "codegen": "npm -s run generate:prisma && npm -s run generate:nexus",
    "generate:prisma": "prisma generate",
    "generate:nexus": "ts-node --transpile-only src/schema",
    "dev": "ts-node-dev --no-notify --respawn --transpile-only src/server",
    "clean": "rimraf build",
    "test": "npm run generate && jest"
  },

```

Im doing this on digital ocean app platform. im getting this error though

Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import i

**Reply by Yasir5247** [reply]
my folder structure follows

```
root
->node_modules
->prisma
->src
->tests
->.env
->.gitIgnore
->DockerFile
->package.json
->tsconfig.json
....

**Reply by Yasir5247** [reply]
this is the exact error shown in the activity console of digital ocean

```
[2022-02-24 14:42:09] 
[2022-02-24 14:42:09] > mcq-api@1.0.0 dev /app
[2022-02-24 14:42:09] > ts-node-dev --no-notify --respawn --transpile-only src/server
[2022-02-24 14:42:09] 
[2022-02-24 14:42:10] [INFO] 14:42:10 ts-node-dev ver. 1.1.8 (using ts-node ver. 9.1.1, typescript ver. 4.5.3)
[2022-02-24 14:42:15] Production API Port is set to: 8080
[2022-02-24 14:42:15] Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
[2022-02-24 14:42:15] In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
[2022-02-24 14:42:15]     at new PrismaClient (/app/node_modules/.prisma/client/index.js:3:11)
[2022-02-24 14:42:15]     at Object.<anonymous> (/app/src/server.ts:12:23)
[2022-02-24 14:42:15]     at Module._compile (internal/modules/cjs/loader.js:999:30)
[2022-02-24 14:42:15]     at Module._compile (/app/node_modules/source-map-support/source-map-support.js:568:25)
[2022-02-24 14:42:15]     at Module.m._compile (/tmp/ts-node-dev-hook-1637127755859411.js:69:33)
[2022-02-24 14:42:15]     at Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
[2022-02-24 14:42:15]     at require.extensions..jsx.require.extensions..js (/tmp/ts-node-dev-hook-1637127755859411.js:114:20)
[2022-02-24 14:42:15]     at require.extensions.<computed> (/tmp/ts-node-dev-hook-1637127755859411.js:71:20)
[2022-02-24 14:42:15]     at Object.nodeDevHook [as .ts] (/app/node_modules/ts-node-dev/lib/hook.js:63:13)
[2022-02-24 14:42:15]     at Module.load (internal/modules/cjs/loader.js:863:32)
[2022-02-24 14:42:15] [ERROR] 14:42:15 Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
[2022-02-24 14:42:15] In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
[]

```

**Reply by Yasir5247** [reply]
I have figured the problem.

RUN npm install --only=production

i'm doing a production install here. so in this case i will not have prisma cli avaibale to run prisma commands such as prisma generate and deploy etc. 

Thanks for the awesome tutorial.

**Comment by wallace-camarinha** [comment]
Amazing content there!
Thanks for sharing this amazing tutorial.

**Comment by raiyansarker** [comment]
Well, I tried the same way, but I am having error in the build step. This is the output,
```
 > [builder 8/8] RUN yarn build:                                                
#14 0.326 yarn run v1.22.19                                                     
#14 0.354 $ tsc -p .                                                            
#14 9.478 src/Post/post.constroller.ts(120,13): error TS7006: Parameter 'error' implicitly has an 'any' type.
#14 9.478 src/auth/createOrUpdateUser.ts(4,10): error TS2305: Module '"@prisma/client"' has no exported member 'Provider'.
#14 9.479 src/auth/createOrUpdateUser.ts(4,20): error TS2305: Module '"@prisma/client"' has no exported member 'Role'.
#14 9.479 src/auth/index.ts(6,10): error TS2305: Module '"@prisma/client"' has no exported member 'Provider'.
#14 9.479 src/auth/sign.ts(4,10): error TS2305: Module '"@prisma/client"' has no exported member 'Role'.
#14 9.479 src/helper/auth/anyUser.ts(1,10): error TS2305: Module '"@prisma/client"' has no exported member 'Role'.
#14 9.479 src/helper/auth/authenticatedUser.ts(1,10): error TS2305: Module '"@prisma/client"' has no exported member 'Role'.
#14 9.479 src/helper/auth/authorOrAdminUser.ts(1,10): error TS2305: Module '"@prisma/client"' has no exported member 'Role'.
#14 9.479 src/helper/auth/authorUser.ts(1,10): error TS2305: Module '"@prisma/client"' has no exported member 'Role'.
#14 9.666 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
#14 9.666 error Command failed with exit code 2.
```

I think it is due to prisma generate not working properly, cause the build works outside of docker. Is there any solution, or am I missing any point?

**Reply by marcjulian** [reply]
Perhaps you have to add an extra step to trigger prisma to generate the client.

You could something like this:

```docker
...
RUN npm install

RUN npx prisma generate
...
```

**Reply by raiyansarker** [reply]
I found the issue, I was using node:16-slim for build and node:16-alpine for production. I am running docker on Macbook, so ARM based architecture systems are having problem running in these images, I don't know why but the binary is not accessible. I saw this post on [StackOverflow](https://stackoverflow.com/questions/66530969/docker-compose-up-issue-using-docker-apple-m1), so I changed both build and production image to node:16, and now everything is working fine.

But the image size is big, I haven't tested with any other image other than the specified above. If anyone have a better solution let me know!

Although there is another issue of prisma not finding the database connector environment variable. I am working on it! BTW, thanks for the article. It really helped a lot.

**Reply by marcjulian** [reply]
I use node:16-alpine for build and production without any issues yet. Maybe it was because of the node:16-slim version. 
Also keep in mind that prisma downloads the prisma-engine for the architecture where you install `prisma`. Probably the architecture for slim and alpine are not the same.

Yes the image is rather larger with node:16. node:16-alpine does a better job but is still big.

Hope you figure it out with the environment variable for prisma! I am glad you like the article!

**Comment by fergalmoran** [comment]
Thank you my friend - clear, concise and works!

**Comment by Kh3an** [comment]
Thanks for the turorial !!
But i ran into an error after the cmd
prisma migrate deploy && npm run start
Error: Can't write to /app/node_modules/@prisma/engines please make sure you install "prisma" with the right permissions.

any idea?

**Reply by marcjulian** [reply]
Haven't seen this error before. How does your dockerfile looks like? maybe it could be an issue with the base node image you are using.

**Comment by ridwansameer** [comment]
This is great! Exactly what was perplexing me because i wasn't aware CMD runs only when the container is run and not built, which is when I want to run the prisma migration!

I do have a question though, in this instance, what is the purpose of the multi-stage build? Could it not be just done in one stage as follows?


```
FROM node:14

# Create app directory
WORKDIR /app

# A wildcard is used to ensure both package.json AND package-lock.json are copied
COPY package*.json ./
COPY prisma ./prisma/

# Install app dependencies
RUN npm install

COPY . .

RUN npm run build

EXPOSE 3000
# 👇 new migrate and start app script
CMD [  "npm", "run", "start:migrate:prod" ]
```

Also wouldn't the `COPY . .` Take care of `COPY prisma ./prisma` since we don't need it during npm install?

Thank you so much :)

**Reply by marcjulian** [reply]
Yes multi stage is not required. Basically up to you which approach is better for your use case.

Yes you are correct `COPY . . ` would also copy the `prisma` directory. Separating the copy steps is helping to build efficient docker images, read the [explanation](https://github.com/notiz-dev/notiz/discussions/169#discussioncomment-1253466) here.

**Comment by vpaivag** [comment]
Nice work! 

I have a question about what can be done in case the migration fails? Because in this way the container would stop when the migration fails. If the migration manages to migrate one part but the other part fails, how can we do a rollback?

**Reply by adrianmanduc** [reply]
Hello! Did you find any solution for this scenario ?

**Comment by mvittiglio** [comment]
I appreciate the article but this method 
1. requires that migrations be already run elsewhere (not really the end of the world but does become yet another thing to manage).
2. has a small bug where `prisma migrate...` isn't found in the container. 

Instead, I added a command to the docker-compose.yml that does the `command: sh -c "npx prisma migrate dev --skip-generate --name 'init' && node server.js"`. I also call generate in an intermediary step in the Dockerfile.

Here is the error I was seeing going with the method documented above: 
```
demo_app-1  | 
demo_app-1  | > demo@0.1.0 demo:migrate
demo_app-1  | > prisma migrate deploy && npm run demo
demo_app-1  | 
demo_app-1  | sh: prisma: not found
demo_app-1 exited with code 127
```

**Comment by marcus13371337** [comment]
A problem with this approach (It's the best I've come up with), if the migration takes a while, a lot of dockerized platforms will try to restart the image if the server isn't up and running after X seconds (responds to a /ping-endpoint). I've run in to cases where our platform will try to restart it in the middle of a migration, and then prisma won't be able to mark the migration as finished automatically, since the container that initialised the migration is shut down. Do you have any ideas on how you could solve such problem?

**Comment by ankita-p17** [comment]
Hello,
This is the only article I could find for what I was looking for. But not able to get the result as you mentioned.
I have generated an initial migration script and manually created new migrations and applied them. Now while deploying with docker and running this. Container gets created but migration is not applied to the database. I'm using node:16-slim and the other steps are the same.
Could you please help here?

**Reply by marcjulian** [reply]
Are you copying the migration files into the docker container? Do you have `prisma` cli installed in your docker container? 
Please provide your Dockerfile or even better a repo for reproduction.

**Reply by ankita-p17** [reply]
I'm using the same dockerfile as above. prisma migration files are getting copied into container asa COPY prisma ./prisma/ this gets executed from above

**Comment by Spoutnik97** [comment]
Would you use the same strategy if the docker image was deployed on a serverless container ? 
I guess it increase the coldstart time

If no, then what would your strategy be ?

**Reply by marcjulian** [reply]
Yes it does add to the coldstart time and also the prisma cli adds to the docker container size.

Perhaps following ["expand and contract pattern"](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern) (Video example https://www.youtube.com/watch?v=5nuHjlKJdFk) could be a possible strategy for your use case.

Effectively you would deploy database migrations, which are **backwards** compatible before deploying the new serverless container.

**Comment by rhufsky** [comment]
Hi, this is a great tutorial., thanks a lot!
One question arises in my case:

When i start npx prisma migrate, obviously prisma must be installed. Normally, in package.json prisma exists in devDependencies and therefore is not available in a production build. Did you move prisma to "dependencies"?

**Comment by davilos** [comment]
nice work!
